packages/models-tree/src/models-tree/ModelsTreeDefinition.ts

```diff
@@ -517,6 +471,143 @@ export class ModelsTreeDefinition implements HierarchyDefinition {
+
+  private async *querySubjects(): AsyncIterableIterator<{ id: Id64String; parentId?: Id64String; targetPartitionId?: Id64String; hideInHierarchy: boolean }> {
+    const subjectsQuery = `
+      SELECT
+        ECInstanceId id,
+        Parent.Id parentId,
+        json_extract(JsonProperties, '$.Subject.Model.TargetPartition') targetPartitionId,
+        CASE
+          WHEN (
+            json_extract(JsonProperties, '$.Subject.Job.Bridge') IS NOT NULL
+            OR json_extract(JsonProperties, '$.Subject.Model.Type') = 'Hierarchy'
+          ) THEN 1
+          ELSE 0
+        END hideInHierarchy
+      FROM bis.Subject
+    `;
+    for await (const row of this._queryExecutor.createQueryReader({ ecsql: subjectsQuery }, { rowFormat: "ECSqlPropertyNames" })) {
+      yield { id: row.id, parentId: row.parentId, targetPartitionId: row.targetPartitionId, hideInHierarchy: !!row.hideInHierarchy };
+    }
+  }
+
+  private async *queryModels(): AsyncIterableIterator<{ id: Id64String; parentId: Id64String }> {
+    const modelsQuery = `
+      SELECT p.ECInstanceId id, p.Parent.Id parentId
+      FROM bis.InformationPartitionElement p
+      INNER JOIN bis.GeometricModel3d m ON m.ModeledElement.Id = p.ECInstanceId
+      WHERE NOT m.IsPrivate AND EXISTS (SELECT 1 FROM bis.GeometricElement3d WHERE Model.Id = m.ECInstanceId)
+    `;
+    for await (const row of this._queryExecutor.createQueryReader({ ecsql: modelsQuery }, { rowFormat: "ECSqlPropertyNames" })) {
+      yield { id: row.id, parentId: row.parentId };
+    }
+  }
+
+  private async initSubjectModels() {
+    this._subjectsHierarchy = new Map();
+    this._subjectModels = new Map();
+    this._subjectInfos = new Map();
+    const targetPartitionSubjects = new Map<Id64String, Set<Id64String>>();
+    await Promise.all([
+      (async () => {
+        for await (const subject of this.querySubjects()) {
+          this._subjectInfos!.set(subject.id, { hideInHierarchy: subject.hideInHierarchy });
+          if (subject.parentId) {
+            pushToMap(this._subjectsHierarchy!, subject.parentId, subject.id);
+          }
+          if (subject.targetPartitionId) {
+            pushToMap(targetPartitionSubjects, subject.targetPartitionId, subject.id);
+          }
+        }
+      })(),
+      (async () => {
+        for await (const model of this.queryModels()) {
+          pushToMap(targetPartitionSubjects, model.id, model.parentId);
+        }
+      })(),
+    ]);
+    for (const [partitionId, subjectIds] of targetPartitionSubjects) {
+      subjectIds.forEach((subjectId) => {
+        pushToMap(this._subjectModels!, subjectId, partitionId);
+      });
+    }
+  }
+
+  private async initCache() {
+    if (!this._init) {
+      this._init = this.initSubjectModels().then(() => {});
+    }
+    return this._init;
+  }
+
+  private forEachChildSubject(parentSubjectId: Id64String, cb: (childSubjectId: Id64String) => "break" | "continue") {
```

Comment:
Maybe using RxJs with `expand()` would be easier?

Reply:
didn't want to introduce `rxjs` dependency just for this